optuna/integration/allennlp/_train.py

```diff
@@ -0,0 +1,378 @@
+            primary_port = (
+                distributed_params.pop("primary_port", None) or common_util.find_open_port()
+            )
+        else:
+            # Otherwise we require that the port be specified.
+            primary_port = distributed_params.pop("primary_port")
+
+        num_procs = len(device_ids)
+        world_size = num_nodes * num_procs
+
+        # Creating `Vocabulary` objects from workers could be problematic since
+        # the data loaders in each worker will yield only `rank` specific
+        # instances. Hence it is safe to construct the vocabulary and write it
+        # to disk before initializing the distributed context. The workers will
+        # load the vocabulary from the path specified.
+        vocab_dir = os.path.join(serialization_dir, "vocabulary")
+        if recover:
+            vocab = Vocabulary.from_files(vocab_dir)
+        else:
+            vocab = training_util.make_vocab_from_params(
+                params.duplicate(), serialization_dir, print_statistics=dry_run
+            )
+        params["vocabulary"] = {
+            "type": "from_files",
+            "directory": vocab_dir,
+            "padding_token": vocab._padding_token,
+            "oov_token": vocab._oov_token,
+        }
+
+        logging.info(
+            "Switching to distributed training mode since multiple GPUs are configured | "
+            f"Primary is at: {primary_addr}:{primary_port} | Rank of this node: {node_rank} | "
+            f"Number of workers in this node: {num_procs} | Number of nodes: {num_nodes} | "
+            f"World size: {world_size}"
+        )
+
+        assert trial is not None
+
+        if isinstance(trial.study._storage, InMemoryStorage):
+            message = "InMemoryStorage is not supported in distributed configuration."
+            message += " You have to use RDB or Redis to use Optuna with AllenNLP distributed."
+            raise ValueError(message)
+
+        elif isinstance(trial.study._storage, _CachedStorage):
+            # Reconstruct storage to purge a old cache.
+            if isinstance(trial.study._storage, _CachedStorage):
+                trial.study._storage = _CachedStorage(trial.study._storage._backend)
+
+        try:
+            mp.spawn(
+                _train_worker_with_optuna,
+                args=(
+                    params.duplicate(),
+                    serialization_dir,
+                    include_package,
+                    dry_run,
+                    node_rank,
+                    primary_addr,
+                    primary_port,
+                    world_size,
+                    device_ids,
+                    file_friendly_logging,
+                    include_in_archive,
+                    trial,
+                ),
+                nprocs=num_procs,
+            )
+        except Exception as e:
+            if "optuna.exceptions.TrialPruned" in str(e):
```

Comment:
(I don't think it is a good way)

When a pruning happens, the worker terminates with `torch.multiprocessing.spawn.ProcessRaisedException`.
To prune a trial in Optuna, it has to raise `optuna.exceptions.TrialPruned` instead of `ProcessRaisedException`.
In the first implementation, I check a traceback and re-raise `TrialPruned` if it is needed.